bfvmm/integration/arch/intel_x64/test_all_inheritance.cpp

```diff
@@ -0,0 +1,98 @@
+//
+// Copyright (C) 2019 Assured Information Security, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// TIDY_EXCLUSION=-cert-err58-cpp
+//
+// Reason:
+//     This test triggers on the use of a std::mutex being globally defined
+//     from the EPT map.
+//
+
+#include <vmm.h>
+
+// -----------------------------------------------------------------------------
+// vCPU
+// -----------------------------------------------------------------------------
+
+ept::mmap g_guest_map;
+
+void
+init()
+{
+    ept::identity_map(g_guest_map, MAX_PHYS_ADDR);
+}
+
+namespace test
+{
+
+class vcpu : public bfvmm::intel_x64::vcpu
+{
+public:
+    explicit vcpu(vcpuid::type id) :
+        bfvmm::intel_x64::vcpu{id}
+    {
+        this->add_external_interrupt_handler(
+            external_interrupt_handler_delegate_t::create<vcpu, &vcpu::test_handler>(this)
+        );
+
+        this->set_eptp(g_guest_map);
+    }
+
+    ~vcpu() override = default;
+
+    bool
+    test_handler(
+        vcpu_t *vcpu, external_interrupt_handler::info_t &info)
+    {
+        vcpu->queue_external_interrupt(info.vector);
+        return true;
+    }
+
+public:
+
+    /// @cond
+
+    vcpu(vcpu &&) = delete;
+    vcpu &operator=(vcpu &&) = delete;
+
+    vcpu(const vcpu &) = delete;
+    vcpu &operator=(const vcpu &) = delete;
+
+    /// @endcond
+};
+
+}
+
+// -----------------------------------------------------------------------------
+// vCPU Factory
+// -----------------------------------------------------------------------------
+
+namespace bfvmm
+{
+
+std::unique_ptr<vcpu>
+vcpu_factory::make(vcpuid::type vcpuid, bfobject *obj)
+{
+    bfignored(obj);
+    return std::make_unique<test::vcpu>(vcpuid);
+}
```

Comment:
We don't. Inheritance is 100% optional with this new interface. My goal was simply to still allow it to be used so that we were not affecting downstream users, integration tests or Boxy. Basically, the new interface is a second, more "C" way of using the APIs.